Contest_API.md

```diff
@@ -252,17 +252,19 @@ absolute timestamps.
     with properties as defined below.
   - Arrays (type **`array of <type>`** in the specification) are built-in JSON 
     arrays of some type defined above.
+  - Nullable types (type **`<type> ?`** in the specification) are either a
+    value of a type defined above, or `null`.
```

Comment:
> * To me 'null' is a valid value for most types, e.g. you can have a null string.

According to the JSON spec a "string" *clearly* can't be null. OTOH, the JSON spec also does not talk about types *at all*. Specifically it does not define a "boolean" but rather the values "true" and "false".

According to JSON Schema, null is its own type, and thus no other type can take the value null. Interestingly (?) it *does* define a boolean type.

> * We've said before that a null value is equivalent to not specifying a property, so to me this is saying that these properties can be null, but isn't saying that others can't be. i.e. it's not really saying anything.

True, but now when we have `/access` there is an important difference... maybe? A value can certainly be missing because it's simply not listed in the access array, but if it *is* and it's not nullable then it can't be null, and therefore it must be there. I.e. a non-nullable value can only be missing if `/access` says so. A nullable value can be missing regardless.

> 'd rather flip this around and specify the required properties, e.g. "Required properties are specified with *, these must exist and be non-null."

I don't *like* it (which is completely subjective) and I think it's not consistent with JSON and JSON Schema (which is at least somewhat objective).

Reply:
I think Tim is using java too much, where every object type can also be null 😅 .
I agree that null should be its own type, but I also see the advantage of using `*`. However, I'm fine with the `?` syntax.

Reply:
Yeah, I probably do use Java too much. 😛

I still think what we have is a set of properties that must exist in every response (i.e. `*`), and the rest are optional. The fact that this implies the rest can be null (except booleans) is secondary IMHO and why I'd rather focus on the required properties.

Reply:
A couple days later and I'm still thinking about this as required properties. As a server implementor I want to know what properties I must support and always return from endpoints; as a client I want to know which properties will always be there. I'm happy to be overruled, but I still think we should flip it.

Reply:
I have to say reading all the ? didn’t make me very happy either. I know I said I didn’t care but actually reading it changed my opinion a bit

Reply:
> As a server implementor I want to know what properties I must support and always return from endpoints; as a client I want to know which properties will always be there. I'm happy to be overruled, but I still think we should flip it.

To me these are two separate questions.

Starting with the second one (i.e. "we should flip if"), I don't think it makes sense at all to say that all data types can be null, but we have a special type operator that makes a type non-nullable. It is much more sane to do it as is in the PR now.

The second question is about what minimum set of properties must always be there. This (arguably) has nothing to do with nullable types. As currently written in this PR the minimum set is empty (which admittedly isn't so useful for clients), but you are told exactly what properties you will get in the access endpoint. I'm on the fence regarding increasing the minimum set to something actually useful, but if we do I would suggest specifying it in the access endpoint.

Reply:
Yes, you're right. These often overlap but are independent, and also how we ended up with two columns before. :-/  In my head I was assuming that was the required column we were effectively retaining, but you've really been doing the nullable.

I do think it's important to keep some representation of required/minimum. It's now impossible to tell which clarification properties I must implement vs which are optional, or whether I need to implement contest start_time (b/c hey, it's nullable). I don't want to bring back a column, but IMHO it is much better to do this inside each endpoint (i.e. in the 'type' definition) vs merged into a common access endpoint.